x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -149,640 +159,455 @@ public JwtRealm(
-            final Environment environment = realmConfig.env();
-            final Path directoryPath = environment.configFile();
-            final Path filePath = directoryPath.resolve(jwkSetPath);
-            // Check file exists, accessible, and non-empty. TODO Maybe return parsed contents here instead of later in a watcher thread.
-            final String fileContents = Files.readString(filePath, StandardCharsets.UTF_8);
-            if (Strings.hasText(fileContents) == false) {
-                throw new Exception(
-                    "Empty file ["
-                        + jwkSetPath
-                        + "] for setting ["
-                        + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
-                        + "]."
-                );
-            }
-            return new Tuple<>(null, filePath); // RETURN URL AS NULL AND PATH AS NON-NULL (i.e. URL Exception is discarded)
-        } catch (Exception e) {
-            pathException = new Exception(
-                "Error loading local file ["
-                    + jwkSetPath
-                    + "] for setting ["
-                    + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
-                    + "].",
-                e
-            );
+    public void ensureInitialized() {
+        if (this.delegatedAuthorizationSupport == null) {
+            throw new IllegalStateException("Realm has not been initialized");
         }
-        // Throw SettingsException with the two suppressed exceptions.
-        final SettingsException settingsException = new SettingsException(
-            "Invalid value [" + jwkSetPath + "] for setting " + RealmSettings.getFullSettingKey(realmConfig, JwtRealmSettings.JWKSET_PATH)
-        );
-        settingsException.addSuppressed(urlException);
-        settingsException.addSuppressed(pathException);
-        throw settingsException;
     }
 
     @Override
     public void initialize(final Iterable<Realm> allRealms, final XPackLicenseState xpackLicenseState) {
+        LOGGER.trace("Initializing realm [" + super.name() + "]");
         if (this.delegatedAuthorizationSupport != null) {
-            throw new IllegalStateException("Realm has already been initialized");
+            throw new IllegalStateException("Realm " + super.name() + " has already been initialized");
         }
         // extract list of realms referenced by super.config.settings() value for DelegatedAuthorizationSettings.AUTHZ_REALMS
         this.delegatedAuthorizationSupport = new DelegatedAuthorizationSupport(allRealms, super.config, xpackLicenseState);
     }
 
-    private void ensureInitialized() {
-        if (this.delegatedAuthorizationSupport == null) {
-            throw new IllegalStateException("Realm has not been initialized");
+    @Override
+    public void close() {
```

Comment:
Close is called by tests, not just node shutdown. I found OIDC and SAML tests which also close the realm before continuing to the next test. Logging close failures is important information to diagnose CI failures.

Logging close failures is also important for support issues. I worked on cases where close() threw an exception, the server got into a bad state, and there was no way to know close() caused that chain reaction. I have even had to deliver dev patches to add missing close() exception warnings, to diagnose those types of issues.

Reply:
Logging again! 

> Close is called by tests, not just node shutdown. I found OIDC and SAML tests which also close the realm before continuing to the next test. Logging close failures is important information to diagnose CI failures.

We can add ad-hoc logging if we have CI failures that we need to investigate. I don't believe that by default being too verbose in our logging just in case it might end up being useful to diagnose a CI failure at some point is worth doing. 

> Logging close failures is also important for support issues. I worked on cases where close() threw an exception, the server got into a bad state, and there was no way to know close() caused that chain reaction. I have even had to deliver dev patches to add missing close() exception warnings, to diagnose those types of issues.

My argument here is that node is closed so there is nothing to be left in a bad state. But my argument might be wrong.

Reply:
It was not an "if" situation. I wrote tests, and the tests randomly failed because I forgot to call close. When I add more tests in later PRs, these logs will tell me if I closed the realm. I could add it later, but it is so minor I am inclined to leave it in.

Reply:
If there was a way for the realm to make an assertion about forgetting to close the realm, I would do that instead, but I can't think of a way to do that.

Actually, if the Realm parent class implemented Closeable, that might be the better solution, but that kind of refactoring is better left to a separate PR.

Reply:
> It was not an "if" situation. I wrote tests, and the tests randomly failed because I forgot to call close. When I add more tests in later PRs, these logs will tell me if I closed the realm. I could add it later, but it is so minor I am inclined to leave it in.

My argument here is that we shouldn't be persisting logs to investigate CI failures we get during development cycles. I find this superfluous and not adding value to the code and this is why I commented. You can add temporary logging while developing, figure out why the test was failing and remove it before merging.

This is getting to the nitpicking category though so I won't push any further, my arguments are all stated above, happy eitherway if you want to resolve this !

Reply:
Removed some logs. Not sure if I got all of them. Not worth effort to debate it so I removed it.